File: advanced-programming-methods/toy-interpreter-java/src/model/expression/Expression.java
package model.expression;

import exceptions.ExpressionException;
import model.IDictionary;
import model.IHeap;

/**
 * The type Expression.
 */
public abstract class Expression implements java.io.Serializable
{
    /**
     * Eval int.
     *
     * @param table the table
     * @param heap  the heap
     * @return the int
     * @throws ExpressionException the expression exception
     */
    abstract public int eval(IDictionary table, IHeap heap) throws ExpressionException;

    abstract public String toString();

    /**
     * Clone deep.
     *
     * @return the expression
     */
    abstract public Expression cloneDeep();
}

File: advanced-programming-methods/labs/toy-interpreter-sharp/model/expression/Expression.cs
using System;
using Exceptions;

namespace Model
{
    /// <summary>
    /// The type Expression.
    /// </summary>
    [Serializable()]
    public abstract class Expression
    {
        /// <summary>
        /// Eval int.
        /// </summary>
        /// <param name="table"> the table </param>
        /// <returns> the int </returns>
        /// <exception cref="ExpressionException"> the expression exception </exception>
        public abstract int Eval(IDictionary table, IHeap heap);

        public override abstract string ToString();

        /// <summary>
        /// Clone deep.
        /// </summary>
        /// <returns> the expression </returns>
        public abstract Expression CloneDeep();
    }
}

